Opshin Language Analysis Report (#1)
Anastasia-Labs/opshin-audit: catalyst_docs/milestone2/milestone2_langanalysis.md

all Python features are available in OpShin.

OpShin ensures that contracts evaluate on-chain exactly as their Python
counterpart. OpShin's compiler ensures that if a program successfully compiles,
it meets two criterias. First, the source code is guaranteed to be a valid
Python program. Second, It ensures the output running it with python is the same
as running it on-chain.

### Limitations

The OpShin language is a subset of python, having the following limitations:

- User-defined symbols can only be imported using `from <pkg> import *`. `import <pkg>` isn't supported.
- Mutual recursion isn't supported
- Classes can't inherit
- Tuples can't contain heterogenous types
- Containers can't contain function values
- Compiler errors are throw immediately when encountered instead of being collected
- ...

### Deviations from python

The limitations of OpShin don't invalidate the claim that it is a subset of python. OpShin however deviates slightly from python, making it not strictly a subset of python:

- ...

## Type System

One of the limitations of using Python as-is for smart contract development is
that it is dynamically typed. The type system of OpShin is much stricter than

> the type system of Python. OpShin addresses this by introducing a strict type

Author: Rewritten as "One limitation of using standard Python for smart contract development is its dynamic typing, which can lead to runtime errors and unpredictability. To solve this, OpShin enforces a strict type system, ensuring type safety and reliability at compile time"

15. Explicitly casts anything that should be boolean (eg. if-then-else conditions, comparison bin ops) by injecting `bool()`
16. Sets the `orig_name` property of Name, FunctionDef and ClassDef AST nodes.
17. Gives each variable a unique name by appending a scope id.
18. Aggressive Type Inference: Visits each AST node to determine its type, setting its `.typ` property in the process.
19. Turns empty lists into raw _pluthon_ expressions.
20. Turns empty dicts into raw _pluthon_ expressions.
21. Ensures that a function that is decorated with a single amed decorated, is decorated with the `@wraps_builtin` decorator, which must be imported using `from opshin.bridge import wraps_builtin`. Such decorated functions are then converted into raw _pluthon_ expressions.
22. Injects the `bytes()`, `int()`, `bool()` and `str()` cast builtins.
23. Removes assignments of types, classes and polymorphic functions (eg. `MyList = List[int]`)
24. (Optional) Iteratively collects all used variables and removes the unused variables. The iteration is stopped if the set of remaining used variables remains unchanged.
25. (Optional) Removes constant statements.
26. Removes Pass AST nodes.

> 27. Generates the _pluthon_ AST

Author: all comments on "." noted and resolved.

# Quantitative Metrics

## Metrics using Gastronomy

> We analysed the `UPLC` code generated by OpShin for a sample validator which

Author: Modified to `We analyzed the _UPLC_ code generated by OpShin for a sample validator that increments an input by 1, using `Gastronomy` as the _UPLC_ debugger. This example highlights some concerns we identified, although it was just one of several tests we conducted to evaluate the compiler’s behavior.`

opshin/std/math.py                                    24      0      8      0   100%
opshin/type_impls.py                                 754     83    495     75    84%
opshin/type_inference.py                             811     35    373     20    94%
opshin/typed_ast.py                                  113      1      0      0    99%
opshin/util.py                                       192     19     60      2    87%
------------------------------------------------------------------------------------
TOTAL                                               4374    471   1802    152    87%

```

# Manual Review Findings

The document herein is provided as an interim update detailing the findings of
our ongoing audit process on the OpShin repository. It is crucial to understand
that this document does not constitute the final audit report. The contents are
meant to offer a preliminary insight into our findings up to this point and are
subject to change as our audit progresses.

## Summary of Current Findings Across Categories

1. Security - 0
2. Performance - 2 (01, 05)
3. Maintainability - 3 (02, 03, 04)

> 4. Others - 0

Author: updated

## Introduction to OpShin

OpShin is a programming language for developing smart contracts on the Cardano blockchain.
It's syntax is 100% valid Python code and it aims to lower the barrier to entry for Cardano smart contract development.
This approach allows developers to write smart contracts using familiar Python constructs while still benefiting from Cardano's robust and secure execution environment.

OpShin ensures that contracts evaluate on-chain exactly as their Python counterpart.
This allows unit tests and verification of the Python code using standard tooling available for Python development.
The type system of opshin is much stricter than the type system of Python, so that many optimizations can be implemented and an elevated level of security is provided.

## Syntax Analysis

OpShin presents itself as a restricted version of Python, written specifically for smart contract development on the Cardano blockchain.
While it encourages developers to write code as they would in standard Python programs, it's important to note that not all Python features are available in OpShin.

OpShin's approach ensures that if a program successfully compiles, it meets two crucial criteria.

> Firstly, the source code is guaranteed to be a valid Python program.
> Secondly, OpShin ensures that the output of running the compiled program with Python is identical to its execution on-chain.

Reviewer: ```suggestion
First, the source code is guaranteed to be a valid Python program.
Second, OpShin ensures that the output of running the compiled program with Python is identical to its execution on-chain.
```

## Type System

One of the limitations of using Python as-is for smart contract development is that it is dynamically typed.
Opshin addresses this concern by introducing a strict type system on top of Python.
What Opshin does is have an independent component called the 'aggressive static type inferencer', which can infer all types of the Python AST nodes for a well chosen subset of Python.

> So in simple terms every variable in OpShin has a type. There are no opaque types in OpShin, everything can be deconstructed.

Reviewer: ```suggestion
So in simple terms every variable in OpShin has a type.
There are no opaque types in OpShin, everything can be deconstructed.
```

Currently, OpShin supports only Lists and Dicts.
It does not support tuples and generic types, which we see as a limitation, as these can be really valuable when writing smart contracts.
This limitation of not supporting tuples and generic types might require workarounds to achieve the desired functionality.

## Compilation and Execution

> OpShin provides a toolkit to evaluate the script in Python,compile the script to UPLC, and compile the script to `pluto`, an intermediate language for debugging purposes.

Reviewer: ```suggestion
OpShin provides a toolkit to evaluate the script in Python, compile the script to UPLC, and compile the script to `pluto`, an intermediate language for debugging purposes.
```

                ]
              ]
            )
            (delay
              (lam
                n_1
                [
                  [
                    (lam
                      1self
                      (lam 1other [ [ (builtin addInteger) 1self ] 1other ])
                    )
                    (force n_1)
                  ]
                  (con integer 1)
                ]
              )
            )
          ]
        )
        (delay
          [
            (lam _ (error ))
            [
              [ (force (builtin trace)) (con string "NameError: validator") ]
              (con unit ())
            ]
          ]
        )
      ]
    )
    (delay
      [
        (lam _ (error ))
        [
          [ (force (builtin trace)) (con string "NameError: n") ] (con unit ())
        ]
      ]
    )
  ]
)
```

The two lambda functions with parameter names `n_1` and `validator_0` correspond to the variables named `n` and `validator`, respectively.
These variables, however, are not being used.
This behavior is expected by the OpShin code, which assumes that it will always be able to access these variables without generating invalid UPLC.

> 

Reviewer: ```suggestion
Also, this has the effect that script sizes are usually bigger then strictly necessary, since dead code elimination is not implemented by the OpShin optimizer.

```

# Code Coverage Percentage

Currently, the unit tests primarily focuses on the builtins, hash library, operations, keywords, ledger functionalities and standard library functions.
The Unit tests that is available covers majority of the codebase.

OpShin's codebase includes two additional modules, `opshin.rewrite` and `opshin.optimize`, which play crucial roles in enhancing the smart contract compiling process.
The `opshin.rewrite` module performs various rewriting steps that remove complexity from Python source code.
Meanwhile, the `opshin.optimize` module ensures better performance through targeted optimizations.
Given the importance of these modules in the OpShin ecosystem, we strongly recommend expanding the unit test coverage to include them.

# Manual Review Findings

The document herein is provided as an interim update detailing the findings of our ongoing audit process on the opshin repository.
It is crucial to understand that this document does not constitute the final audit report.
The contents are meant to offer a preliminary insight into our findings up to this point and are subject to change as our
audit progresses.

> ## Summary of Findings Across Categories

Reviewer: ```suggestion
## Summary of Current Findings Across Categories
```

> 1. Security - Nil
> 2. Performance - Nil
> 3. Maintainability - 4
> 4. Others - Nil

Reviewer: ```suggestion
1. Security - 0
2. Performance - 0
3. Maintainability - 4
4. Others - 0
```
